doc/source/signals.rst

```diff
@@ -0,0 +1,168 @@
+  def send_signal(value):
+      signal.send(UserSignal(value))
+
+.. autofunction:: ray.experimental.signal.receive
+
+Here is a simple example of how to receive signals from an actor or task identified
+by ``a``. Note that an actor is identified by its handle, and a task by one of its
+object ID return values.
+
+.. code-block:: python
+
+  import ray.experimental.signal as signal
+
+  # This returns a possibly empty list of all signals that have been sent by 'a'
+  # since the last invocation of signal.receive from within this process. If 'a'
+  # did not send any signals, then this will wait for up to 10 seconds to receive
+  # a signal from 'a'.
+  signal_list = signal.receive([a], timeout=10)
+
+.. autofunction:: ray.experimental.signal.reset
+
+
+Example: sending a user signal
+------------------------------
+
+The code below show a simple example in which a task, called ``send_signal()``
+sends a user signal and the driver gets it by invoking ``signal.receive()``.
+
+.. code-block:: python
+
+  import ray.experimental.signal as signal
+
+  # Define a user signal.
+  class UserSignal(signal.Signal):
+      def __init__(self, value):
+            self.value = value
+
+      def get_value(self):
+            return self.value
+
+  @ray.remote
+  def send_signal(value):
+      signal.send(UserSignal(value))
+      return
+
+  signal_value = 'simple signal'
+  object_id = send_signal.remote(signal_value)
+  # Wait up to 10sec to receive a signal from the task. Note the task is
+  # identified by the object_id it returns.
+  result_list = signal.receive([object_id], timeout=10)
+  # Print signal values. This should print "simple_signal".
+  # Note that result_list[0] is the signal we expect from the task.
+  # The signal is a tuple where the first element is the first object ID
+  # returned by the task and the second element is the signal object.
+  print(result_list[0][1].get_value())
+
+Example: Getting an error signals
+---------------------------------
+
+This is a simple example in which a driver gets an error signal caused
+by the failure of ``task()``.
+
+.. code-block:: python
+
+  @ray.remote
+  def task():
+      raise Exception('exception message')
+
+  object_id = task.remote()
+  try:
+      ray.get(object_id)
+  except Exception as e:
+      pass
+  finally:
+      result_list = signal.receive([object_id], timeout=10)
+      # Expected signal is 'ErrorSignal'.
+      assert type(result_list[0][1]) == signal.ErrorSignal
+      # Print the error.
+      print(result_list[0][1].get_error())
+
+
+Example: Sending signals between multiple actors
+------------------------------------------------
+
+This is a more involved example in which two actors ``a1`` and ``a2`` each
+generate five signals, and another actor ``b`` waits to receive all signals
+generated by ``a1`` and ``a2``, respectively. Note that ``b`` recursively calls
+its own method ``get_signals()`` until it gets all signals it expects.
+
+.. code-block:: python
+
```

Comment:
Modify this example to be self-contained. It's not clear what imports need to be imported or how many CPUs we need to run this.